### PR TITLE
AndroidChromeStrategy: Always unformat input on keydown

### PR DESCRIPTION
**Pull request by intelliot**
With third party keyboards on Android, the `keypress` event never occurs. We still get the `keydown` event, but currently, we only unformat if the event `_isDeletion`.

This removes the `_isDeletion` check and always unformats on `keydown`.

Keyboards:
- [x] Google keyboard
- [x] Samsung keyboard
- [x] TouchPal
- [ ] ~Sony Xperia (pressing delete does not activate auto formatting)~ Wasn't able to find on Google Play store
- [x] Swype
- [x] SwiftKey Keyboard

Other actions:
- [x] Autofill
- [x] Pastes - they don't work with the Samsung `Clipboard` feature, but normal pastes appear to be fine

### CODE REVIEW

File: lib/strategies/android-chrome.js [hunk 1]
```diff
@@ -9,6 +9,30 @@ function AndroidChromeStrategy(options) {
...
+  });
+  self.inputElement.addEventListener('keypress', function (event) {
+    if (keyCannotMutateValue(event)) { return; }
+    self._unformatInput(event);
+  });
+  self.inputElement.addEventListener('keyup', function (event) {
+    self._reformatInput(event);
+  });
+  self.inputElement.addEventListener('input', function (event) {
+    // Safari AutoFill fires CustomEvents -- mark the input as unformatted without actually unformatting the current value
```
Comment (crookedneighbor) [inline lib/strategies/android-chrome.js]: Safari?

Comment (intelliot) [inline lib/strategies/android-chrome.js]: Copied from `base.js`

Comment (crookedneighbor) [inline lib/strategies/android-chrome.js]: I'm curious if Android Chrome fires the same event or not.

Comment (intelliot) [inline lib/strategies/android-chrome.js]: I tried Android Chrome with the Google keyboard and the Samsung keyboard. I put a `console.log` inside the `if` block. It was not called. Shall I remove this bit of code, or should we leave it "just in case"?

Comment (intelliot) [inline lib/strategies/android-chrome.js]: On Android Chrome, autofill isn't an option unless I add the `autocomplete` attribute to the Hosted Fields. But if I do, then autofill seems to work fine (although it will only do one field at a time, likely due to the iframes).

Comment (crookedneighbor) [inline lib/strategies/android-chrome.js]: I understand that for our purposes, we only care about hosted fields, but this library can be used outside of hosted fields and I'd be interested to know about cc autofill behavior in a normal form.

Comment (intelliot) [inline lib/strategies/android-chrome.js]: I tested the cc autofill behavior in a normal form on Android Chrome. It worked. No `CustomEvent`s were fired.

**Comment by crookedneighbor** [comment]
Would like to try this out myself, but I _think_ it looks good.

File: lib/strategies/android-chrome.js [hunk 2]
```diff
@@ -9,6 +10,26 @@ function AndroidChromeStrategy(options) {
 AndroidChromeStrategy.prototype = Object.create(BaseStrategy.prototype);
 AndroidChromeStrategy.prototype.constructor = AndroidChromeStrategy;
 
+BaseStrategy.prototype._attachListeners = function () {
```
Comment (EvanHahn) [inline lib/strategies/android-chrome.js]: Could we shorten this to:

```js
var self = this;
var keyUpAndInputHandler = self._reformatInput.bind(self);

function keyDownAndKeyPressHandler(event) {
  if (keyCannotMutateValue(event)) { return; }
  self._unformatInput(event);
}

self.inputElement.addEventListener('keydown', keyDownAndKeyPressHandler);
self.inputElement.addEventListener('keypress', keyDownAndKeyPressHandler);

self.inputElement.addEventListener('keyup', keyUpAndInputHandler);
self.inputElement.addEventListener('input', keyUpAndInputHandler);

self.inputElement.addEventListener('paste', self.pasteEventHandler.bind(self));
```

It's a little less repetitive, though it may not be as clear. Thoughts?

Comment (crookedneighbor) [inline lib/strategies/android-chrome.js]: Also, you're overwriting the BaseInput prototype, not the AndroidChrome Strategy prototype

Comment (crookedneighbor) [inline lib/strategies/android-chrome.js]: I'm happy with evan's suggestion.

**Comment by crookedneighbor** [comment]
Other than my one comment about not overwriting the BaseInput prototype, I think this looks good.

File: lib/strategies/android-chrome.js (hunk 2 above)
Comment (crookedneighbor) [inline lib/strategies/android-chrome.js]: Mixed feelings about this.

I'm slightly :+1: about the DRY version, but it does make the code harder to parse at a glance and if something slight needs to change with the keydown event, we'll need to undry it anyway.

¯\_(ツ)_/¯

**Comment by lilaconlee** [comment]
🐈
